[MRG] Refactor `feature_selection.f_regression` and introduce `feature_selection.r_regression`

jjerphan opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Follow-up of  #8353
Fixes #7195


#### What does this implement/fix? Explain your changes.

Refactors `feature_selection.f_regression` for convenience: `feature_selection.f_regression` now relies on `feature_selection.r_regression`, a public function which pre-computes Pearson's R

#### Other remarks

`feature_selection.abs_r_regression`, introduced in #8353, which returned the absolute values of `feature_selection.r_regression` was initially considered.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

DSLituiev committed:
r_regression and abs_r_regression added

DSLituiev committed:
r_regression and abs_r_regression added

DSLituiev committed:
indentation fix

DSLituiev committed:
whitespace fix

DSLituiev committed:
import and tests

DSLituiev committed:
indentation

DSLituiev committed:
code style

DSLituiev committed:
init fix

DSLituiev committed:
Merge branch 'master' into r_regression

DSLituiev committed:
Merge branch 'master' into r_regression

jjerphan committed:
Merge branch 'master' into r_regression

jjerphan committed:
Change assert_true for assert

jjerphan committed:
Change module for import to match current one

I made this change to have the test ran as I previously got:

```
sklearn/feature_selection/tests/test_feature_select.py:None (sklearn/feature_selection/tests/test_feature_select.py)
/home/jsquared/.virtualenvs/sk/lib64/python3.8/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
../__init__.py:27: in <module>
    from ._mutual_info import mutual_info_regression, mutual_info_classif
../_mutual_info.py:9: in <module>
    from ..neighbors import NearestNeighbors
../../neighbors/__init__.py:17: in <module>
    from ._nca import NeighborhoodComponentsAnalysis
../../neighbors/_nca.py:22: in <module>
    from ..decomposition import PCA
../../decomposition/__init__.py:17: in <module>
    from .dict_learning import dict_learning
E   ModuleNotFoundError: No module named
'sklearn.decomposition.dict_learning'
```

It seems that there are reason for this special handling
to exist according to the comment above.

This might need to be reverted.

jjerphan committed:
Add reference in doc comment

jjerphan committed:
Revert "Change module for import to match current one"

This reverts commit 63d6fd759bcf3acadbc5cb1d6a144523a7d60e09.

jjerphan commented:
In fact, it looks like this PR is just a really small re-factoring to have two new coefficients accessible.
@jnothman: regarding what you've said in the previous PR (#8353), is it still relevant to benchmark it?

jjerphan committed:
Merge branch 'master' into r_regression

jjerphan committed:
Merge branch 'master' into r_regression

jjerphan commented:
What do you think, @jnothman? :)

jjerphan commented:
Hi @jnothman, this is a friendly reminder for this stalled PR follow-up.
Do you think that it is still worth benchmarking it?

cmarmo commented:
Hi @jjerphan thanks for your pull request. There are linting issues and a conflict that should be solved before review. Also if you think that this is ready for review, do you mind changing the title from [WIP] to [MRG]? Thanks a lot for your patience!

jjerphan committed:
Merge branch 'main' into r_regression

jjerphan committed:
Merge branch 'main' into r_regression

jjerphan commented:
@cmarmo: are you interested in reviewing this PR? 🙂

cmarmo commented:
Hi @jjerphan, thanks for your patience! The documentation for the two new functions is not generated: this is because a reference to the function should be added in the [doc/modules/classes.rst](https://github.com/scikit-learn/scikit-learn/blob/main/doc/modules/classes.rst) file.
To bring more attention to your pull request I believe it could be a good idea to propose the requested benchmarks: scikit-learn contains a lot of lines of code, adding more features adds maintenance costs. Nice benchmarks provide more motivation to feature addition. Thanks for your collaboration.

jjerphan commented:
Hi @cmarmo, thanks for your feedback.

> The documentation for the two new functions is not generated: this is because a reference to the function should be added in the doc/modules/classes.rst file.

I will do it now.

> To bring more attention to your pull request I believe it could be a good idea to propose the requested benchmarks: scikit-learn contains a lot of lines of code, adding more features adds maintenance costs. Nice benchmarks provide more motivation to feature addition.

Thanks for confirming this. I will have a look at this too.

jjerphan committed:
Add documentation reference for abs_r_regression and r_regression

jjerphan committed:
Complete test to include r_regression

jjerphan commented:
Here is the output of [a simple script for the benchmark](https://gist.github.com/jjerphan/0cd9adf82fe795e66e5d492da0be85bb) to compare the running time of `SelectKBest` under with `abs_r_regression` and `f_regression`.

```
Dataset-shape: (10000, 10) {'scorer': <function f_regression at 0x7fd2fd0539d0>} 	 100 trials: 0.0028 ± 0.0011 s (median = 0.0027 s)
Dataset-shape: (10000, 10) {'scorer': <function abs_r_regression at 0x7fd2fd053940>} 	 100 trials: 0.0026 ± 0.0012 s (median = 0.0023 s)

Dataset-shape: (10000, 100) {'scorer': <function f_regression at 0x7fd2fd0539d0>} 	 100 trials: 0.0100 ± 0.0007 s (median = 0.0099 s)
Dataset-shape: (10000, 100) {'scorer': <function abs_r_regression at 0x7fd2fd053940>} 	 100 trials: 0.0093 ± 0.0008 s (median = 0.0092 s)

Dataset-shape: (10000, 1000) {'scorer': <function f_regression at 0x7fd2fd0539d0>} 	 100 trials: 0.0666 ± 0.0034 s (median = 0.0665 s)
Dataset-shape: (10000, 1000) {'scorer': <function abs_r_regression at 0x7fd2fd053940>} 	 100 trials: 0.0646 ± 0.0030 s (median = 0.0643 s)

Dataset-shape: (100000, 10) {'scorer': <function f_regression at 0x7fd2fd0539d0>} 	 100 trials: 0.0211 ± 0.0013 s (median = 0.0208 s)
Dataset-shape: (100000, 10) {'scorer': <function abs_r_regression at 0x7fd2fd053940>} 	 100 trials: 0.0207 ± 0.0008 s (median = 0.0205 s)

Dataset-shape: (100000, 100) {'scorer': <function f_regression at 0x7fd2fd0539d0>} 	 100 trials: 0.0833 ± 0.0076 s (median = 0.0824 s)
Dataset-shape: (100000, 100) {'scorer': <function abs_r_regression at 0x7fd2fd053940>} 	 100 trials: 0.0807 ± 0.0031 s (median = 0.0802 s)

Dataset-shape: (100000, 1000) {'scorer': <function f_regression at 0x7fd2fd0539d0>} 	 100 trials: 0.4069 ± 0.0042 s (median = 0.4069 s)
Dataset-shape: (100000, 1000) {'scorer': <function abs_r_regression at 0x7fd2fd053940>}  100 trials: 0.4046 ± 0.0033 s (median = 0.4043 s)
```

`SelectKBest` with `abs_r_regression` has a slightly shorter running time.
I don't think that it is really significant.

#### Benchmark environment 

- Python: 3.9.1
- OS: Fedora 33
- CPU: i7-4600U CPU @ 2.10GHz (4VCores)
- RAM: 8GB

cmarmo commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -230,19 +230,16 @@ def chi2(X, y):
-    1. The correlation between each regressor and the target is computed,
-       that is, ((X[:, i] - mean(X[:, i])) * (y - mean_y)) / (std(X[:, i]) *
+    The cross correlation between each regressor and the target is computed,
+    that is, ((X[:, i] - mean(X[:, i])) * (y - mean_y)) / (std(X[:, i]) *
        std(y)).
```
```suggestion
    that is, ((X[:, i] - mean(X[:, i])) * (y - mean_y)) / (std(X[:, i]) *
    std(y)).
```
This should fix the sphinx warning.

jjerphan committed:
Remove Sphinx warning on indentation

Co-authored-by: Chiara Marmo <cmarmo@users.noreply.github.com>

ogrisel commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -141,7 +141,7 @@ def f_classif(X, y):
         The set of p-values.
 
-    See Also
+    See also
```
Why change this? This is the convention in scikit-learn:

```
(dev) ogrisel@mba scikit-learn % git grep "See Also" | wc -l
     265
(dev) ogrisel@mba scikit-learn % git grep "See also" | wc -l
      33
```

https://scikit-learn.org/stable/developers/contributing.html#guidelines-for-writing-documentation

jjerphan replied:
Ah yes, I hesitated when merging `main` in.
I'll revert that.

ogrisel commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -204,7 +204,7 @@ def chi2(X, y):
     Complexity of this algorithm is O(n_classes * n_features).
 
-    See Also
+    See also
```
same comment here.

ogrisel commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -259,22 +256,21 @@ def f_regression(X, y, *, center=True):
+        Pearson R correlation coefficients of features.
 
-    See Also
+    See also
```
again

ogrisel commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -230,19 +230,16 @@ def chi2(X, y):
+    The cross correlation between each regressor and the target is computed,
+    that is, ((X[:, i] - mean(X[:, i])) * (y - mean_y)) / (std(X[:, i]) *
+    std(y)).
 
```
Maybe add a comments such as:

`r_regression` can be used to rank features as a slightly cheaper alternative to `f_regression` if p-values are not needed. Note however that `r_regression` values can be positive or negative.

`abs_r_regression` can be used to rank features by correlation magnitude instead.

jjerphan replied:
I've added a such comments in the docstring of `f_regression` accordingly.
I guess `abs_r_regression` is the main alternative to it for variable selection.

I also have slightly rectified the docstring of `r_regression`

ogrisel commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +296,75 @@ def f_regression(X, y, *, center=True):
+    that require a statistic which is increases with significance of
+    association.
+
+    See also
```
```suggestion
    See Also
```

ogrisel commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +296,75 @@ def f_regression(X, y, *, center=True):
+    pval : array, shape=(n_features,)
+        p-values of F-scores.
+
+    See also
```
```suggestion
    See Also
```

ogrisel commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +296,75 @@ def f_regression(X, y, *, center=True):
+
+    See also
+    --------
+    r_regression: Univariate linear regression tests returning Pearson R.
```
Maybe also link to abs_r_regression for feature selection with `SelectKBest`.

ogrisel commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -259,22 +256,21 @@ def f_regression(X, y, *, center=True):
-    SelectFpr : Select features based on a false positive rate test.
-    SelectFdr : Select features based on an estimated false discovery rate.
-    SelectFwe : Select features based on family-wise error rate.
-    SelectPercentile : Select features based on percentile of the highest
```
Mention `abs_r_regression`.

ogrisel commented on sklearn/feature_selection/tests/test_feature_select.py:
```diff
@@ -71,6 +72,25 @@ def test_f_classif():
+    pearson_r_sparse = coeff(sparse.csr_matrix(X), y, center=False)
+    assert_array_almost_equal(pearson_r_sparse, pearson_r)
+
+
```
Please test the equivalence with the ordering returned by `f_regression`'s F statistics, preferably on a a wide dataset (e.g. with n_samples=200 and n_features=1000) where feature selection is typically used:

```python
assert_array_equal(f_regression(X, y)[0].argsort(),
                   abs_r_regression(X, y).argsort())
```

and please also check:

Please add a test to check that:

```python
assert_allclose(abs_r_regression(X, y), np.abs(r_regression(X, y)))
```

possibly in the same test, e.g. named `test_f_regression_r_regression_consistency`.

ogrisel reviewed (approved):
LGTM: just a few suggestion to revert the change on "See Also" to uniformly use the title case convention which we currently follow for that section in scikit-learn (for some reason). I did not comment on all occurrences. See the diff tab in github to catch them all.

Thanks for the benchmark, indeed the computational difference is minimal. Still I think it's good to have those for educational purpose. I did not realize that the F statistics resulted in the same order as the absolute value of the cross-correlation coefficient.

Other than this, one additional test and we are good to go.

jjerphan committed:
Use 'See Also' over 'See Also'

See Guidelines for writing documentation:

https://scikit-learn.org/stable/developers/contributing.html#guidelines-for-writing-documentation

Co-authored-by: Olivier Grisel <olivier.grisel@gmail.com>

jjerphan committed:
Clarify docstrings

jjerphan committed:
Add consistency test

jjerphan committed:
Test for Pearson R correct support

Also use a better wording.

jjerphan committed:
fixup! Clarify docstrings

cmarmo commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -230,19 +230,17 @@ def chi2(X, y):
+def r_regression(X, y, *, center=True):
+    """Compute Pearson R correlation coefficients between the features and
+    the target
 
```
```suggestion

    .. versionadded:: 1.0

```

jjerphan replied:
Good catch! 

The indentation looks correct. I've looked at others contributions adding new interfaces for the 1.0: they used the `versionadded` at different places in the docstrings and some used `1.0.0` for the version number.

What should I use exactly?

cmarmo replied:
I think 1.0 is good: I can't recall versionadded directives for bug fix versions...

cmarmo commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +293,92 @@ def f_regression(X, y, *, center=True):
 
+def abs_r_regression(X, y, center=True):
+    """Absolute value of Pearson R from univariate linear regressions.
+
```
```suggestion

    .. versionadded:: 1.0

```

cmarmo reviewed (approved):
Thanks @jjerphan , I'm not sure about the indentation in my suggestions. It's better if you test them locally... :)
This and a change log entry for version 1.0 then LGTM!

jjerphan committed:
Merge branch 'main' into r_regression

jjerphan committed:
Add docstring directives for additions in version 1.0

Co-authored-by: Chiara Marmo <chiara.marmo@u-psud.fr>

cmarmo reviewed (approved):
Thanks @jjerphan , perhaps @glemaitre might have a look?

jjerphan commented:
Gentle up @glemaitre. 🙂

glemaitre commented:
I put in my TODO but I am draining my backlog first :)

glemaitre commented on doc/whats_new/v1.0.rst:
```diff
@@ -103,6 +103,17 @@ Changelog
+:mod:`sklearn.feature_selection`
+................................
+
+- |Feature| :func:`feature_selection.abs_r_regression` is new criterion
```
```suggestion
- |Feature| :func:`feature_selection.abs_r_regression` is a new criterion
```

glemaitre commented on sklearn/feature_selection/__init__.py:
```diff
@@ -44,6 +46,8 @@
            'f_oneway',
            'f_regression',
+           'r_regression',
+           'abs_r_regression',
```
Could you move this line above `chi2`

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -230,19 +230,19 @@ def chi2(X, y):
 
 
 @_deprecate_positional_args
```
You can remove this decorator because we introduce a new function.

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -230,19 +230,19 @@ def chi2(X, y):
-    """Univariate linear regression tests.
+def r_regression(X, y, *, center=True):
+    """Compute Pearson R correlation coefficients between the features and
+    the target
```
```suggestion
    the target.
```

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -230,19 +230,19 @@ def chi2(X, y):
-def f_regression(X, y, *, center=True):
-    """Univariate linear regression tests.
+def r_regression(X, y, *, center=True):
+    """Compute Pearson R correlation coefficients between the features and
```
```suggestion
    """Compute Pearson's R correlation coefficients between the features and
```

glemaitre commented on doc/whats_new/v1.0.rst:
```diff
@@ -103,6 +103,17 @@ Changelog
+- |Feature| :func:`feature_selection.abs_r_regression` is new criterion
+  which can be used with :class:`feature_selection.SelectKBest` to select
+  variables. It is the absolution values of
+  :func:`feature_selection.r_regression` which computes Pearson R correlation
```
```suggestion
  :func:`feature_selection.r_regression` which computes Pearson's R correlation
```

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -230,19 +230,19 @@ def chi2(X, y):
-       that is, ((X[:, i] - mean(X[:, i])) * (y - mean_y)) / (std(X[:, i]) *
-       std(y)).
-    2. It is converted to an F score then to a p-value.
+    The cross correlation between each regressor and the target is computed,
+    that is, ((X[:, i] - mean(X[:, i])) * (y - mean_y)) / (std(X[:, i]) *
+    std(y)).
```
```suggestion
    The cross correlation between each regressor and the target is computed
    as ((X[:, i] - mean(X[:, i])) * (y - mean_y)) / (std(X[:, i]) * std(y)).
```

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -259,23 +259,18 @@ def f_regression(X, y, *, center=True):
-
-    pval : array, shape=(n_features,)
-        p-values of F-scores.
+    corr : array, shape=(n_features,)
```
Since the function is simple, I would suggest to change the code to call `corr` -> `corr_coef` (or even `correlation_coefficient`)

```suggestion
    corr_coef : ndarray of shape (n_features,)
```

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -259,23 +259,18 @@ def f_regression(X, y, *, center=True):
-
-    pval : array, shape=(n_features,)
-        p-values of F-scores.
+    corr : array, shape=(n_features,)
```
Could you modify the docstring above with:

```
    X : {array-like, sparse matrix}  of shape (n_samples, n_features)
        The set of regressors being tested sequentially.
    y : array-like of shape (n_samples,)
        The target vector.
    center : bool, default=True
        Whether or not to center the data matrix `X`. By default, `X`
        will be centered.
```

glemaitre replied:
Could you update similarly the docstring of `f_regression` then.

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +295,94 @@ def f_regression(X, y, *, center=True):
     corr = safe_sparse_dot(y, X)
```
you can remove the comment `# compute the correlation`. The code should be self explicit if the name of the variable is explicit :)

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +295,94 @@ def f_regression(X, y, *, center=True):
+       std(y)) using r_regression function.
+    2. It is converted to an F score and then to a p-value.
+
+    If p-values are not needed, r_regression can be used to rank features as
```
```suggestion
    If p-values are not needed, :func:`r_regression` can be used to rank features as
```

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +295,94 @@ def f_regression(X, y, *, center=True):
+    2. It is converted to an F score and then to a p-value.
+
+    If p-values are not needed, r_regression can be used to rank features as
+    a slightly cheaper alternative to f_regression. Note however that
```
```suggestion
    a slightly cheaper alternative to :func:`f_regression`. Note however that
```

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +295,94 @@ def f_regression(X, y, *, center=True):
+
+    If p-values are not needed, r_regression can be used to rank features as
+    a slightly cheaper alternative to f_regression. Note however that
+    contrary to f_regression, r_regression values lie in [-1, 1] and can thus
```
```suggestion
    contrary to :func:`f_regression`, :func:`r_regression` values lie in [-1, 1] and can thus
```

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +295,94 @@ def f_regression(X, y, *, center=True):
+    contrary to f_regression, r_regression values lie in [-1, 1] and can thus
+    be negative.
+
+    Alternatively, `abs_r_regression` can be used to rank features by
```
```suggestion
    Alternatively, :func:`abs_r_regression` can be used to rank features by
```

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +295,94 @@ def f_regression(X, y, *, center=True):
+
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}  shape = (n_samples, n_features)
+        The set of regressors that will be tested sequentially.
+
+    y : array of shape(n_samples).
+        The data matrix
+
+    center : True, bool,
+        If true, X and y will be centered.
+
+    Returns
+    -------
+    F : array, shape=(n_features,)
+        F values of features.
+
+    pval : array, shape=(n_features,)
+        p-values of F-scores.
```
```suggestion
    X : {array-like, sparse matrix}  shape = (n_samples, n_features)
        The set of regressors being tested sequentially.

    y : array-like of shape (n_samples,)
        The target vector.

    center : True, bool,
        Whether or not to center the data matrix `X`. By default, `X`
        will be centered.

    Returns
    -------
    f_statistic : ndarray of shape (n_features,)
        F-statistic for each feature.

    p_values : ndarray of shape (n_features,)
        P-values associated with the F-statistic.
```

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +295,94 @@ def f_regression(X, y, *, center=True):
+        scores.
+    """
+
+    # compute the correlation
+    corr = r_regression(X, y, center=center)
```
```suggestion
    correlation_coefficient = r_regression(X, y, center=center)
```

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +295,94 @@ def f_regression(X, y, *, center=True):
+    # compute the correlation
+    corr = r_regression(X, y, center=center)
     degrees_of_freedom = y.size - (2 if center else 1)
+    # convert to p-value
```
```suggestion
        corr_coef_squared = correlation_coefficient ** 2
        f_statistic = corr_coef_squared / (1 - corr_coef_squared) * degrees_of_freedom
        p_values = stats.f.sf(f_statistic, 1, degrees_of_freedom)
```

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +295,94 @@ def f_regression(X, y, *, center=True):
 
 
+def abs_r_regression(X, y, center=True):
+    """Absolute value of Pearson R from univariate linear regressions.
```
```suggestion
    """Absolute value of Pearson's R from univariate linear regressions.
```

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +295,94 @@ def f_regression(X, y, *, center=True):
+
+    .. versionadded:: 1.0
+
+    This convenience wrapper is to be used with SelectKBest and other models
+    that require a statistic which is increased with significance of
+    association.
```
```suggestion
    This wrapper is convenient when using :class:`~sklearn.feature_selection.SelectKBest`
    and other models that require a statistic which is increased with significance of
    association.
```

I am unsure of the meaning of `significance of association` thought (it might be right ;))

jjerphan replied:
> I am unsure of the meaning of significance of association thought (it might be right ;))

So do I, to be honest.

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +295,94 @@ def f_regression(X, y, *, center=True):
+
+    See Also
+    --------
+    r_regression: Univariate linear regression tests returning Pearson R.
```
```suggestion
    r_regression: Univariate linear regression tests returning Pearson's R
        correlation coefficients.
```

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +295,94 @@ def f_regression(X, y, *, center=True):
+    r_regression: Univariate linear regression tests returning Pearson R.
+    SelectKBest: Select features based on the k highest scores.
+    """
+    # compute the correlation
```
```suggestion
```

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -300,14 +295,94 @@ def f_regression(X, y, *, center=True):
+    SelectKBest: Select features based on the k highest scores.
+    """
+    # compute the correlation
+    corr = r_regression(X, y, center=center)
+    return abs(corr)
```
```suggestion
    return np.abs(r_regression(X, y, center=center))
```

glemaitre commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -502,12 +577,14 @@ class SelectKBest(_BaseFilter):
+    f_classif: ANOVA F-value between label/feature for classification tasks.
+    mutual_info_classif: Mutual information for a discrete target.
+    chi2: Chi-squared stats of non-negative features for classification tasks.
+    abs_r_regression: Absolute value of Pearson R between label and features
```
```suggestion
    abs_r_regression: Absolute value of Pearson's R correlation coefficient between label and features
```

glemaitre commented on sklearn/feature_selection/tests/test_feature_select.py:
```diff
@@ -18,9 +19,10 @@
-    chi2, f_classif, f_oneway, f_regression, mutual_info_classif,
-    mutual_info_regression, SelectPercentile, SelectKBest, SelectFpr,
-    SelectFdr, SelectFwe, GenericUnivariateSelect)
+    chi2, f_classif, f_oneway, f_regression, abs_r_regression, r_regression,
```
since we are modifying these imports, could make an import by line in a `black` style:

```python
from sklearn.feature_selection import (
    chi2,
    f_classif,
...
)

glemaitre commented on sklearn/feature_selection/tests/test_feature_select.py:
```diff
@@ -71,6 +73,27 @@ def test_f_classif():
+    # with centering, compare with sparse
+    correlation_coeffs = coeff(X, y, center=True)
+    correlation_coeffs_sparse = coeff(sparse.csr_matrix(X), y, center=True)
+    assert_array_almost_equal(correlation_coeffs_sparse, correlation_coeffs)
```
c

glemaitre commented on sklearn/feature_selection/tests/test_feature_select.py:
```diff
@@ -71,6 +73,27 @@ def test_f_classif():
+    # again without centering, compare with sparse
+    correlation_coeffs = coeff(X, y, center=False)
+    correlation_coeffs_sparse = coeff(sparse.csr_matrix(X), y, center=False)
+    assert_array_almost_equal(correlation_coeffs_sparse, correlation_coeffs)
```
could you use `assert_allclose` instead.

glemaitre commented on sklearn/feature_selection/tests/test_feature_select.py:
```diff
@@ -71,6 +73,27 @@ def test_f_classif():
 
 
+@pytest.mark.parametrize("coeff", [abs_r_regression, r_regression])
+def test_r_regression(coeff):
```
I would split the test to be more granular:

* one test to check the consistence between sparse and not sparse
* one that check the centering (this could also use `sklearn.testing._convert_container) to test on dense and sparse array
* and one that check the absolute equivalence

glemaitre replied:
Do we want to check that our implementation is close to the `np.corrcoef` from numpy?

glemaitre commented on sklearn/feature_selection/tests/test_feature_select.py:
```diff
@@ -97,6 +120,20 @@ def test_f_regression():
+
+    assert_array_equal(Fs.argsort(), abs_r_regression(X, y).argsort())
+
+    # Test consistency of definition
```
you can remove the comment

glemaitre reviewed (changes requested):
A first review.

jjerphan committed:
Fix typo in the whats_new entry

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
End sentence with a full stop.

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Correct typo in docstring.

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Correct typo in the whats_new entry.

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Fix syntax in r_regression's docstring.

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Sort imports alphabetically

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Remove useless decorator

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Use verbose name for the correlation coefficient

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Improve r_regression and r_regression docstring

Use accurate descriptions for the parameters.

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Improve code comments

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Add Sphinx domains in f_regression's docstring

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Make f statistics' and their p-values' computations clearer

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Improve wording abs_r_regression

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Use black formatting style for imports

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Remove useless comments

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
Prefer assert_allclose over assert_array_almost_equal

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan committed:
fixup! Prefer assert_allclose over assert_array_almost_equal

Revert changes for unrelated tests.

thomasjpfan commented:
With this PR, `r_regression` and `abs_r_regression` would be the only functions in `sklearn.feature_selection` that does not return p-values. Given the computational cost of p-values is not high: https://github.com/scikit-learn/scikit-learn/pull/17169#issuecomment-779347745, I see the benefit of this PR is the convenience method for absolute value, which is is shorthand for:


```python
from sklearn.datasets import make_regression
from sklearn.feature_selection import SelectKBest, f_regression
import numpy as np

X, y = make_regression(random_state=0)

f_abs_regression = lambda *args: np.abs(f_regression(*args)[0])
X_new = SelectKBest(f_abs_regression, k=10).fit_transform(X, y)
```

If we want to make `abs` easier to use, we can add a parameter to `SelectKBest`:

```python
X_new = SelectKBest(f_regression, k=10, transform_func=np.abs).fit_transform(X, y)
```

I am wondering if adding 2 new functions worth the maintenance cost for making `SelectKBest` easier to use.

jjerphan committed:
Split test into several

Also check against numpy's implementation of the
correlation coefficient.

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jjerphan commented:
Thanks for the comprehensive review, @glemaitre. I should have adressed all your requests.

IMO, the value of this PR is the introduction of `r_regression`, and I see no problem to remove the originally proposed `abs_r_regression`.

I would go for @thomasjpfan's first suggestion, hence assuming that it is up to users to define the `score_func` they want `SelectKBest` to be used with.

thomasjpfan commented:
`r_regression` is `f_regression` without the p value computation. For consistency, are we add going to add `anova_classif` which is `f_classif` without the p value computation? 

There is also the option of adding a `return_p_value` to `f_regression`, which gives the option to compute the p value.

jjerphan commented:
> `r_regression` is `f_regression` without the p value computation.

`f_regression` relies on `r_regression` but return different statistics.

Still, it would make sense allowing a way to specify the computation of the p-values as you propose.
I don't think that I am the best to know what the most suitable choice here is.

jjerphan committed:
fixup! Improve wording abs_r_regression

glemaitre commented:
I think that we could drop `abs_f_regression`. Then, I still think that we need `r_regression` and `f_regression` as proposed. Additionally, it might be nice to add a keyword `return_p_values` (True by default to be back-compatible) that allows bypassing the computation if p-values are not needed.

glemaitre commented:
So we can limit the scope for this PR to introduce `r_regression` only.

jjerphan committed:
Drop `feature_selection.abs_r_regression`

See discussions:
https://github.com/scikit-learn/scikit-learn/pull/17169#issuecomment-820997349

glemaitre reviewed (approved):
LGTM

ogrisel commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -229,83 +229,131 @@ def chi2(X, y):
-def f_regression(X, y, *, center=True):
-    """Univariate linear regression tests.
+def r_regression(X, y, *, center=True):
+    """Compute Pearson's R correlation coefficients between the features and
+    the target.
```
PEP 257 multi-line docstring for better integration with sphinx / IDEs:

```suggestion
    """Compute Pearson's r for each features and the target.

    Pearson's r is also known as the Pearson correlation coefficient.
```

ogrisel replied:
Also it's lowercase "r" according to Wikipedia.

ogrisel commented on sklearn/feature_selection/_univariate_selection.py:
```diff
@@ -229,83 +229,131 @@ def chi2(X, y):
+       std(y)) using r_regression function.
+    2. It is converted to an F score and then to a p-value.
+
+    If p-values are not needed, :func:`r_regression` can be used to rank
+    features as a slightly cheaper alternative to :func:`f_regression`.
+    Note however that contrary to :func:`f_regression`, :func:`r_regression`
+    values lie in [-1, 1] and can thus be negative.
```
I find this paragraph confusing. Let me try to rephrase:

```suggestion
    :func:`f_regression` is derived from :func:`r_regression` and will rank
    features in the same order if all the features are positively correlated
    with the target.

    Note however that contrary to :func:`f_regression`, :func:`r_regression`
    values lie in [-1, 1] and can thus be negative. :func:`f_regression` is
    therefore recommended as a feature selection criterion to identify
    potentially predictive feature for a downstream classifier, irrespective of
    the sign of the association with the target variable.

    Furthermore :func:`f_regression` returns p-values while
    :func:`r_regression` does not.
```

ogrisel reviewed (approved):
Just a few small improvement suggestions:

jjerphan committed:
Use multi-line docstring (PEP 257) for better integration with IDEs

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Rephrase docstring

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

ogrisel commented:
Merged! Thanks @jjerphan.

jjerphan commented:
'welcome!

Thanks to @DSLituiev for the original PR!
